pkgs/applications/networking/xpipe/default.nix

```
@@ -0,0 +1,71 @@
+{ stdenvNoCC, lib, fetchzip, makeDesktopItem }:
+
+let
+  inherit (stdenvNoCC.hostPlatform) system;
+  throwSystem = throw "Unsupported system: ${system}";
+
+  arch = {
+    x86_64-linux = "x86_64";
+    aarch64-linux = "arm64";
+  }.${system} or throwSystem;
+
+  hash = {
+    x86_64-linux = "sha256-VmJa0DqnFSzq4wJ+HchguDK4tMeKP/XaE2MidXfootM=";
+    aarch64-linux = "sha256-LdXfhZK4n8HqN1u/rKIIZhFk7fREXbTuAzyvvhFR6Lo=";
+  }.${system} or throwSystem;
+
+  displayname = "XPipe";
+
+in stdenvNoCC.mkDerivation rec {
+  pname = "xpipe";
+  version = "1.6.0";
+
+  src = fetchzip {
+    url = "https://github.com/xpipe-io/${pname}/releases/download/${version}/xpipe-portable-linux-${arch}.tar.gz";
+    inherit hash;
+  };
+
+  desktopItem = makeDesktopItem {
+    categories = [ "Network" ];
+    comment = "The shell connection hub and remote file browser for your entire infrastructure";
+    desktopName = "${displayname}";
+    exec = "/opt/${pname}/cli/bin/xpipe open %U";
+    genericName = "Shell connection hub";
+    icon = "/opt/${pname}/logo.png";
+    name = "${displayname}";
+  };
+
+  installPhase = ''
+    runHook preInstall
+
+    pkg="${pname}"
+    mkdir -p $out/opt/$pkg
+    cp -r ./ $out/opt/$pkg
+
+    mkdir -p "$out/usr/bin"
+    ln -s "$out/opt/$pkg/cli/bin/xpipe" "$out/usr/bin/$pkg"
+
+    mkdir -p "$out/usr/share/applications"
+    cp -r "${desktopItem}/share/applications/" "$out/usr/share/"
```

There's a hook called `copyDesktopItems` that will handle installing the desktop item, as long as it is put into `desktopItems`

Note that you'll need to add it to the function up at the top, which is not included in the suggestion diff.

```suggestion
  desktopItems = [
    (makeDesktopItem {
      categories = [ "Network" ];
      comment = "The shell connection hub and remote file browser for your entire infrastructure";
      desktopName = "${displayname}";
      exec = "/opt/${pname}/cli/bin/xpipe open %U";
      genericName = "Shell connection hub";
      icon = "/opt/${pname}/logo.png";
      name = "${displayname}";
    })
  ]

  nativeBuildInputs = [ copyDesktopItems ];

  installPhase = ''
    runHook preInstall
    
    pkg="${pname}"
    mkdir -p $out/opt/$pkg
    cp -r ./ $out/opt/$pkg
    
    mkdir -p "$out/usr/bin"
    ln -s "$out/opt/$pkg/cli/bin/xpipe" "$out/usr/bin/$pkg"
    
    mkdir -p "$out/usr/share/applications"
    cp -r "${desktopItem}/share/applications/" "$out/usr/share/"
``

---

Will I still be able to access the path of the desktop file to use substituteInPlace on it?

---

Oh right! No, not really. But I also see that most packages doesn't do that - they install the icons into `$out/share/icons/hicolor/<resolution>/`, and just specify `icon = "xpipe";` (for `xpipe.png`). Replacing `Exec` is not necessary either, as the binary will exist in the `$PATH` of the user.

Have a look at https://discourse.nixos.org/t/what-is-the-recommended-use-of-makedesktopitem-how-to-setup-the-icon-correctly/13388/6, and maybe grep around nixpkgs a bit to see how others are using `makeDesktopItem`.

---

Your suggested method would definitely look better, but I want to keep the current one with the absolute path to the executable and icon mainly because all other packages like for AUR also do it this way. This approach makes the installation more encapsulated and I can install an xpipe package with multiple package managers on the same system without them interfering too much. (I mainly do this for testing)

---

I won't make this a blocker, but I find this reasoning is a bit weird. The package isn't in nixpkgs for developers to test them with multiple package managers, they are there for normal users to use them. I'd keep them in mind as the target audience when making these decisions, and rather use stuff like overlays and overrides for any special needs.

I also don't really get the encapsulation issues given that we have the nix store which encapsulates stuff very well, but I guess an argument could be made for broken environments.